### PR TITLE
LibGfx: Support for SOF0 JPEG

### PR DESCRIPTION
**Pull request by LucasChollet**
Now you may be asking yourselves: "But Lucas, aren't you supposed to work on SOF2 support, and we already have support for SOF0."
And yes, we have some support but SOF0, but not for weirdos that you never encounter.

The goal of this PR is to add support for reading multiple scans (I know that I claimed that it was straightforward in the last PR :sweat_smile:). In a nutshell, we now put scan-related information in a Scan object in order to have a defined lifetime. It allows us to have control over which component is provided in a scan and support scans with different subsampling.

Even if you never encounter them, one can generate a SOF0 JPEG with each component clustered in a different scan. And we now support them!
Also, if this is very rare for SOF0, having multiple scans is mandatory for SOF2 images. Actually, this PR brings us really close to SOF2 JPEG with only spectral selection.

<details>
<summary>How to generate weirdos with cjpeg</summary>
With a "scan_infos" file containing:

```txt
0: 0 63 0 0;
1: 0 63 0 0;
2: 0 63 0 0;
```

```bash
cjpeg -scans scan_infos input_file > output_file
```

</details>

### CODE REVIEW

File: Userland/Libraries/LibGfx/JPEGLoader.cpp [hunk 1]
```diff
@@ -177,6 +177,14 @@ struct ICCMultiChunkState {
     FixedArray<ByteBuffer> chunks;
 };
 
+struct Scan {
```
Comment (nico) [inline Userland/Libraries/LibGfx/JPEGLoader.cpp:197]: s/class/struct/ in commit description

File: Userland/Libraries/LibGfx/JPEGLoader.cpp [hunk 2]
```diff
@@ -124,13 +124,25 @@ struct MacroblockMeta {
...
 };
 
-struct ComponentSpec {
-    u8 id { 0 };
-    u8 hsample_factor { 1 }; // Horizontal sampling factor.
-    u8 vsample_factor { 1 }; // Vertical sampling factor.
-    u8 ac_destination_id { 0 };
-    u8 dc_destination_id { 0 };
-    u8 qtable_id { 0 }; // Quantization table id.
+/**
```
Comment (nico) [inline Userland/Libraries/LibGfx/JPEGLoader.cpp:127]: use `//` comments

File: Userland/Libraries/LibGfx/JPEGLoader.cpp [hunk 3]
```diff
@@ -136,6 +136,11 @@ struct Component {
     u8 hsample_factor { 1 }; // Hi, Horizontal sampling factor
     u8 vsample_factor { 1 }; // Vi, Vertical sampling factor
     u8 qtable_id { 0 };      // Tqi, Quantization table destination selector
+
+    // The JPEG specification does not specify which component correspond to
```
Comment (nico) [inline Userland/Libraries/LibGfx/JPEGLoader.cpp:140]: s/correspond/corresponds/

Isn't it just in order?

Isn't just `index` a better name maybe?

File: Userland/Libraries/LibGfx/JPEGLoader.cpp [hunk 4]
```diff
@@ -596,6 +596,11 @@ static ErrorOr<void> read_start_of_scan(AK::SeekableStream& stream, JPEGLoadingC
     current_scan.spectral_selection_end = TRY(stream.read_value<u8>());
     current_scan.successive_approximation = TRY(stream.read_value<u8>());
 
+    dbgln_if(JPEG_DEBUG, "Start of Selection: {}, End of Selection: {}, Successive Approximation: {}!",
```
Comment (nico) [inline Userland/Libraries/LibGfx/JPEGLoader.cpp:608]: s/!// maybe? :)

File: Userland/Libraries/LibGfx/JPEGLoader.cpp [hunk 5]
```diff
@@ -201,6 +201,8 @@ struct Scan {
     u8 spectral_selection_start {};
     u8 spectral_selection_end {};
     u8 successive_approximation {};
+
+    HuffmanStreamState huffman_stream;
```
Comment (nico) [inline Userland/Libraries/LibGfx/JPEGLoader.cpp:205]: "Please note that this patch has intentional implicit behavior. The" -> "Please note that this patch changes behavior. The"

File: Userland/Libraries/LibGfx/JPEGLoader.cpp [hunk 6]
```diff
@@ -203,6 +203,12 @@ struct Scan {
     u8 successive_approximation {};
 
     HuffmanStreamState huffman_stream;
+
+    // See the note under the first paragraph
```
Comment (nico) [inline Userland/Libraries/LibGfx/JPEGLoader.cpp:207]: Which first paragraph? What if someone adds a paragraph somewhere? Reword this so it's clear what this refers to.

**Review by nico — APPROVED** [review]
Seems alright, but maybe add a jpeg created with the command in the details and add a test that loads it to Tests/LibGfx/TestImageDecoder.cpp.

Does this have any implications for 4-channel jpegs? (Since you're removing the check for the number of components iirc)

File: Userland/Libraries/LibGfx/JPEGLoader.cpp (hunk 3 above)
Comment (LucasChollet) [reply Userland/Libraries/LibGfx/JPEGLoader.cpp:140]: >Isn't it just in order?

They are ordered, and it's why we are setting it to the index inside the vector. But the idea is to be able to retrieve it from a ScanComponent without going through the whole vector of Components.
Or maybe you're asking if scans are ordered following the components' definition in the start of frame. In this case, the answer is no.

>Isn't just index a better name maybe?

I wanted to make it clear that it is specific to our implementation, and index might be confused with id.

File: Userland/Libraries/LibGfx/JPEGLoader.cpp (hunk 4 above)
Comment (LucasChollet) [reply Userland/Libraries/LibGfx/JPEGLoader.cpp:608]: Leftover from a copy-paste :sweat_smile:

File: Userland/Libraries/LibGfx/JPEGLoader.cpp (hunk 3 above)
Comment (LucasChollet) [reply Userland/Libraries/LibGfx/JPEGLoader.cpp:140]: >I wanted to make it clear that it is specific to our implementation

I added it to the comment

**Comment by LucasChollet** [comment]
As there is no easy way to test correctness in TestImageDecoder, I don't really see the value of adding a test to it right now. But that's definitely something that we should do.

> Does this have any implications for 4-channel jpegs?

We still reject JPEGs with a components count that differs from 1 or 3 at the frame level, the check that I removed is at the scan level. So from the "don't panic the program" point of view, no, but it's now way easier to implement support for it :^). (Basically just detect it and make the CMYK to RGB conversion)

**Comment by nico** [comment]
Why do you say there's no easy way to test correctness? Before, the image you described couldn't be decoded at all. So just doing something like `EXPECT_EQ(plugin_decoder->size(), Gfx::IntSize(417, 223));` (or whatever) should work, no?

**Comment by LucasChollet** [comment]
The decoder succeeding doesn't mean that it produce a correct output, and testing image size isn't sufficient.
If you drop c05b724 the decoder will still succeed, but the image will have artifacts (the vr image on discord).

**Comment by nico** [comment]
But some test coverage (for the dimensions) is probably better than none at all? Also, then the test input is in the repo and can be used for manual testing as well.

(And if you really want to teat contents, you could also check in a PNG of the correctly decoded jpeg and compare to that, i suppose)

**Comment by LucasChollet** [comment]
> But some test coverage (for the dimensions) is probably better than none at all? Also, then the test input is in the repo and can be used for manual testing as well.

Fine, let's do that for now.

> (And if you really want to test contents, you could also check in a PNG of the correctly decoded jpeg and compare to that, i suppose)

I plan to do something like this with a small bitmap, so we can embed the result in the cpp file. And other decoder should do that too, I considered that it's too much work for this PR.

File: Tests/LibGfx/TestImageDecoder.cpp [hunk 7]
```diff
@@ -101,6 +101,23 @@ TEST_CASE(test_jpeg_sof0_one_scan)
     EXPECT(frame.duration == 0);
 }
 
+TEST_CASE(test_jpeg_sof0_several_scans)
+{
+    auto file = MUST(Core::MappedFile::map(TEST_INPUT("several_scans.jpg"sv)));
+    EXPECT(Gfx::JPEGImageDecoderPlugin::sniff(file->bytes()));
+    auto plugin_decoder = MUST(Gfx::JPEGImageDecoderPlugin::create(file->bytes()));
+    EXPECT(plugin_decoder->initialize());
+
```
Comment (nico) [inline Tests/LibGfx/TestImageDecoder.cpp:110]: for bonus points, describe how you created the file in the commit description of this commit

File: Tests/LibGfx/TestImageDecoder.cpp [hunk 8]
```diff
@@ -101,6 +101,23 @@ TEST_CASE(test_jpeg_sof0_one_scan)
...
+    EXPECT(plugin_decoder->frame_count());
+    EXPECT(!plugin_decoder->is_animated());
+    EXPECT(!plugin_decoder->loop_count());
```
Comment (nico) [inline Tests/LibGfx/TestImageDecoder.cpp:113]: no need to check animation state

File: Tests/LibGfx/TestImageDecoder.cpp [hunk 9]
```diff
@@ -101,6 +101,23 @@ TEST_CASE(test_jpeg_sof0_one_scan)
...
+
+    EXPECT(!plugin_decoder->frame(0).is_error());
```
Comment (nico) [inline Tests/LibGfx/TestImageDecoder.cpp:115]: remove

File: Tests/LibGfx/TestImageDecoder.cpp [hunk 10]
```diff
@@ -101,6 +101,23 @@ TEST_CASE(test_jpeg_sof0_one_scan)
...
+
+    auto frame = plugin_decoder->frame(0).release_value_but_fixme_should_propagate_errors();
```
Comment (nico) [inline Tests/LibGfx/TestImageDecoder.cpp:117]: `auto frame = MUST(plugin_decoder->frame(0));`

…but I thought you wanted to check the size, not get the frame? (getting the frame is fine too, also serves in exercising the new code I suppose)

**Review by nico — APPROVED** [review]
Thanks!

The test looks like it's copy-pasted from old other test cases – they got modernized a day or two ago, consider rebasing and using the new tests as pattern instead of the old ones.

File: Tests/LibGfx/TestImageDecoder.cpp (hunk 10 above)
Comment (LucasChollet) [reply Tests/LibGfx/TestImageDecoder.cpp:117]: You were the one proposing testing the size, I don't think either of it has value.

File: Tests/LibGfx/TestImageDecoder.cpp (hunk 7 above)
Comment (LucasChollet) [reply Tests/LibGfx/TestImageDecoder.cpp:110]: You'd better really send me those bonus points :eyes:

**Pull request merged**